AsyncNats/NatsConnection.cs

```diff
@@ -482,63 +544,90 @@ private async IAsyncEnumerable<T> InternalSubscribe<T>(string subject, string? q
                 try
                 {
                     await SendUnsubscribe(subscription);
-                    _subscriptions = _subscriptions.Where(s =>s != subscription).ToArray();
+                    _subscriptions.TryRemove(subscription.SubscriptionId, out _);
                 }
                 finally
                 {
                     _subscriptionsLock.Release();
                 }
             }
         }
+        private T DeserializeWrapper<T>(Func<NatsMsg, T> deserialize, NatsMsg message)
```

Comment:
What's the added benefit of this function? Making InternalSubscribe slightly less complex? Or to supply the 'Unsafe' version? [which I'd like to see removed, not a huge fan of 'Unsafe']

Reply:
I understand your concern. After seeing a few Unsafe* methos in dotnet codebase itself and other libraries I came to terms with it. Yes it can cause memory leaks if not used correctly, but that would be self inflicted and the Unsafe prefix make it clear.
The benefit is being able to inspect and process the message without the extra memory copy.
Another option would be to instead of asking the user to release the NatsMsg, to make clear that the the user can't hold a reference to that instance because it will be released on the next iteration. If the user decides to keep the message, it needs to copy it. 
And yet another option is having a subscribe method that accepts a delegate

Reply:
I've played with the idea of not making a copy of "NatsMsg" in the Subscribe at all and adding a comment that if a reference to NatsMsg is required it's up to the user to actually make a copy (and/or call "Rent").  In all honesty, I think 'raw' Subscribe will hardly be used in real world scenario's (and if it is, a form of the Unsafe version will probably be preferable).

Reply:
You are probably right. I made it based on a particular weird use case of mine, where I will end up discarding most of the messages I'm subscribed to after inspecting headers. I can remove from the PR and leave it on a branch of mine not to pollute the public library.

Reply:
We can keep it, not a huge fan but as you said, other libraries do it too. It gives people options.

Reply:
OK! 
Do you think it's better to ask users to call both Rent() and Release() or just Release() with Rent() being implicit as it is now?

Reply:
Yes, that was what I was thinking as well, the use-case for 'manually' parsing NatsMsg and only using it inside the iteration is larger than also storing the NatsMsg for later use. And not having the payload is better explainable than generating a memory leak (which, depending on load, may be more difficult to spot).

Reply:
Good. I'll make that change. It makes sense.

Reply:
Done.